File: internal/api/contactcache.go
package api

import (
	"net/http"

	"github.com/lichti/zaplab/internal/whatsapp"
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

// getContactCache returns cached contacts, optionally filtered by search query.
func getContactCache(e *core.RequestEvent) error {
	type row struct {
		ID          string `db:"id"               json:"id"`
		JID         string `db:"jid"              json:"jid"`
		Name        string `db:"name"             json:"name"`
		Phone       string `db:"phone"            json:"phone"`
		About       string `db:"about"            json:"about"`
		AvatarURL   string `db:"avatar_url"       json:"avatar_url"`
		LastSeen    string `db:"last_seen"        json:"last_seen"`
		CacheUpdAt  string `db:"cache_updated_at" json:"cache_updated_at"`
	}

	q := pb.DB().
		Select("id", "jid", "name", "phone", "about", "avatar_url", "last_seen", "cache_updated_at").
		From("contact_cache").
		OrderBy("name ASC").
		Limit(500)

	if s := e.Request.URL.Query().Get("q"); s != "" {
		q = q.AndWhere(dbx.NewExp(
			"(name LIKE {:q} OR phone LIKE {:q} OR jid LIKE {:q})",
			dbx.Params{"q": "%" + s + "%"},
		))
	}

	var rows []row
	if err := q.All(&rows); err != nil {
		return e.JSON(http.StatusInternalServerError, map[string]any{"error": err.Error()})
	}
	if rows == nil {
		rows = []row{}
	}
	return e.JSON(http.StatusOK, map[string]any{"contacts": rows, "total": len(rows)})
}

// postContactCacheRefresh forces a live fetch and cache update for a JID.
func postContactCacheRefresh(e *core.RequestEvent) error {
	jidStr := e.Request.URL.Query().Get("jid")
	if jidStr == "" {
		return e.JSON(http.StatusBadRequest, map[string]any{"error": "jid is required"})
	}
	jid, ok := whatsapp.ParseJID(jidStr)
	if !ok {
		return e.JSON(http.StatusBadRequest, map[string]any{"error": "invalid jid"})
	}
	info, err := whatsapp.GetContactInfo(jid)
	if err != nil {
		return e.JSON(http.StatusInternalServerError, map[string]any{"error": err.Error()})
	}
	return e.JSON(http.StatusOK, info)
}

// deleteContactCacheEntry removes a single JID from the cache.
func deleteContactCacheEntry(e *core.RequestEvent) error {
	id := e.Request.PathValue("id")
	record, err := pb.FindRecordById("contact_cache", id)
	if err != nil {
		return e.JSON(http.StatusNotFound, map[string]any{"error": "not found"})
	}
	if err := pb.Delete(record); err != nil {
		return e.JSON(http.StatusInternalServerError, map[string]any{"error": err.Error()})
	}
	return e.JSON(http.StatusOK, map[string]any{"deleted": true})
}

// postContactCachePopulate runs a background job to populate the cache for all known contacts.
func postContactCachePopulate(e *core.RequestEvent) error {
	go func() {
		contacts, err := whatsapp.GetAllContacts()
		if err != nil {
			return
		}
		for _, c := range contacts {
			jid, ok := whatsapp.ParseJID(c.JID)
			if !ok {
				continue
			}
			whatsapp.GetContactInfo(jid) //nolint:errcheck — triggers cache upsert
		}
	}()
	return e.JSON(http.StatusOK, map[string]any{"message": "cache population started in background"})
}

File: pb_public/js/sections/contactcache.js
// Contact Cache — browse cached contact metadata and manage cache population.
function contactCacheSection() {
  return {
    cc: {
      contacts: [],
      total:    0,
      loading:  false,
      error:    '',
      toast:    null,
      search:   '',
    },
    ccRefreshJid: '',

    async ccLoad() {
      this.cc.loading = true;
      this.cc.error   = '';
      try {
        const p = new URLSearchParams();
        if (this.cc.search) p.set('q', this.cc.search);
        const res  = await fetch(`/zaplab/api/contact-cache?${p}`, { headers: apiHeaders() });
        const data = await res.json();
        if (!res.ok) throw new Error(data.error || 'Failed');
        this.cc.contacts = data.contacts || [];
        this.cc.total    = data.total    || 0;
      } catch (err) {
        this.cc.error = err.message;
      } finally {
        this.cc.loading = false;
      }
    },

    async ccPopulate() {
      this.cc.loading = true;
      this.cc.toast   = null;
      try {
        const res  = await fetch('/zaplab/api/contact-cache/populate', {
          method: 'POST', headers: apiHeaders(),
        });
        const data = await res.json();
        if (!res.ok) throw new Error(data.error || 'Failed');
        this.cc.toast = { ok: true, message: 'Cache population started in background' };
        setTimeout(() => this.ccLoad(), 3000);
      } catch (err) {
        this.cc.toast = { ok: false, message: err.message };
      } finally {
        this.cc.loading = false;
      }
    },

    async ccRefresh() {
      if (!this.ccRefreshJid.trim()) return;
      this.cc.loading = true;
      this.cc.toast   = null;
      try {
        const res  = await fetch(`/zaplab/api/contact-cache/refresh?jid=${encodeURIComponent(this.ccRefreshJid.trim())}`, {
          method: 'POST', headers: apiHeaders(),
        });
        const data = await res.json();
        if (!res.ok) throw new Error(data.error || 'Failed');
        this.cc.toast    = { ok: true, message: `Refreshed: ${data.jid}` };
        this.ccRefreshJid = '';
        await this.ccLoad();
      } catch (err) {
        this.cc.toast = { ok: false, message: err.message };
      } finally {
        this.cc.loading = false;
      }
    },

    async ccDelete(id) {
      this.cc.loading = true;
      try {
        const res = await fetch(`/zaplab/api/contact-cache/${id}`, {
          method: 'DELETE', headers: apiHeaders(),
        });
        if (!res.ok) { const d = await res.json(); throw new Error(d.error || 'Failed'); }
        await this.ccLoad();
      } catch (err) {
        this.cc.toast = { ok: false, message: err.message };
      } finally {
        this.cc.loading = false;
      }
    },
  };
}
